pkg/querier/tail.go

```diff
@@ -0,0 +1,92 @@
+package querier
+
+import (
+	"context"
+	"time"
+
+	"github.com/grafana/loki/pkg/iter"
+	"github.com/grafana/loki/pkg/logproto"
+)
+
+// tailIteratorIncrement is for defining size of time window we want to query entries for
+const tailIteratorIncrement = 10 * time.Second
+
+// delayQuerying is for delaying querying of logs for specified seconds to not miss any late entries
+const delayQuerying = 10 * time.Second
+
+func (q *Querier) tailQuery(ctx context.Context, queryStr, regexp string) iter.EntryIterator {
+	return &tailIterator{
+		from:     time.Now().Add(-(tailIteratorIncrement + delayQuerying)),
+		queryStr: queryStr,
+		regexp:   regexp,
+		querier:  q,
+		ctx:      ctx,
+	}
+}
+
+type tailIterator struct {
+	queryStr string
+	regexp   string
+	from     time.Time
+	err      error
+	ctx      context.Context
+	querier  *Querier
+	iter.EntryIterator
```

Comment:
I'd probably avoid embedded this struct like this (instead I'd name it).  Right now you're implementing almost all the methods on EntryIterator on tailIterator, so the embedding doesn't buy you much, and it made lead to confusion in the future.